engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util/errwrap"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+)
+
+const (
+	// HetznerPollLimit sets a lower limit on polling interval in seconds.
+	// Note that the hetzner API supports bursts at up to 3600 requests per hour.
+	// A higher polling interval is recommended when expecting high rates
+	// of operational http requests to the hetzner api.
+	HetznerPollLimit = 1
+	// WaitIntervalLimit sets a lower limit on wait intervals in seconds.
+	// Nonzero values disable burst requests to manage rate limit errors.
+	// A wait interval >= the polling interval is recommended.
+	WaitIntervalLimit = 0
+
+	// StateRunning indicates the server is powered on.
+	StateRunning = "running"
+	// StateOff indicates the server is turned off.
+	StateOff = "off"
+	// StateAbsent indicates the server does not exist at this time.
+	StateAbsent = "absent"
+
+	// ServerRescueDisabled indicates that rescue mode should be disabled.
+	ServerRescueDisabled = "disabled"
+	// ServerRescueTypeLinux32 specifies the image to be used in rescue mode.
+	ServerRescueTypeLinux32 = "linux32"
+	// ServerRescueTypeLinux64 specifies the image to be used in rescue mode.
+	ServerRescueTypeLinux64 = "linux64"
+	// ServerRescueTypeFreeBSD64 specifies the image to be used in rescue mode.
+	ServerRescueTypeFreeBSD64 = "freebsd64"
+)
+
+func init() {
+	engine.RegisterResource("hetzner", func() engine.Res { return &HetznerRes{} })
+}
+
+// HetznerRes is a Hetzner cloud resource. It connects with the cloud API
+// using the hcloud-go package provided by Hetzner. The API token for a new
+// project should be generated manually (eg. via the cloud console) before this
+// resource can establish a connection with the API. One hetzner resource
+// represents one server instance, even if its client has access to multiple
+// instances. A resource in the `absent` state only exists as a local
+// blueprint, and does not exist as server instance on Hetzner's side.
+// Note: for now, all manually registered SSH keys are used when creating a
+// server. More selective use to be added soon.
+type HetznerRes struct {
+	traits.Base
+	init *engine.Init
+
+	// APITokenPath is the path of a file containing the project's API token.
+	// Provide the full path, incl. filetype, and keep the token private.
+	// Consider where the hcloud CLI stores tokens in default operation.
+	APIToken string `lang:"apitoken"`
+	// ServerType determines how much ram, disk and cpu the resource has.
+	Servertype string `lang:"servertype"`
```

Comment:
Good points. To reply to them all in order:
- ServerType instead of Servertype is fine for me, it makes more sense to write it that way. 
- Hetzner does use the "datacenter" spelling in their API, so I think it is most consistent to stick with that?
- list of server options is tricky, see below

The full list of options is only available by querying it from the API. I think this was done so that Hetzner doesn't have to maintain their static documentation when changes are made. I can put the current list of options in the comments somewhere, or as a set of constants, but they can become outdated over time so they will require maintenance. 

Alternatively, I could add an API query to Init() after creating the client, which validates the server options? I wasn't sure if API interactions within Init() are supported, but it's probably the cleanest way of adding a sanity check on those inputs. It doesn't provide the user with any starting info though, unless we add a list of valid options to the error message if an invalid option is specified. 

A more creative solution could be to build a little golang project in an independent repo that requests all server options and dumps them into a file for manual lookups. I'd be happy to host that on my github, and then people can use copies as needed. Then we can just add a link to that repo in the resource documentation for reference. It's not ideal, but at least it solves the problem of people having to write their own scripts to obtain the full options. 

Personally I tend towards the last option since it doesn't overcomplicate the Hetzner resource itself, and offers a little open source contribution for hcloud-specific projects independently of mgmt. Let me know what you prefer.

Reply:
Let's keep it simple for now, but at minimum docstring should have a URL that tells the reader where the list of current values can be found (some link to the hetzner website probably.)

> Hetzner does use the "datacenter" spelling in their API, so I think it is most consistent to stick with that?
Yup, use what they use.

Reply:
Agreed with keeping things simple, but afaik there is no website that lists the current options. Besides my proposed solutions, all I can do is refer users to the API docs and have them collect valid options on their own using the CLI (hcloud) or the API  wrappers for golang (hcloud-go) or python (hcloud-python). 

https://docs.hetzner.cloud

Reply:
Doublechecked the Hetzner docs, and ServerType seems like the correct option. Changed it in the next commit.